src/tools/tidy/src/rustdoc_html_ids.rs

```diff
@@ -0,0 +1,152 @@
+//! * There are no unused IDs.
+
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::fs::File;
+use std::io::{BufRead, BufReader};
+use std::path::Path;
+
+use regex::Regex;
+
+const ID_MAP_PATH: &str = "librustdoc/html/markdown.rs";
+const IDS_USED_IN_JS: &[&str] = &[
+    // This one is created in the JS cannot be found in rust files.
+    "help",
+];
+
+fn extract_ids(path: &Path, bad: &mut bool) -> HashMap<String, usize> {
+    let file = File::open(path).expect("failed to open file to extract rustdoc IDs");
+    let buf_reader = BufReader::new(file);
+    let mut iter = buf_reader.lines();
+    let mut ids = HashMap::new();
+
+    while let Some(Ok(line)) = iter.next() {
+        if line.starts_with("fn init_id_map() -> ") {
+            break;
+        }
+    }
+    // We're now in the function body, time to retrieve the IDs!
+    while let Some(Ok(line)) = iter.next() {
+        if line.trim_start().starts_with("map.insert(\"") {
+            let id = line.split('"').skip(1).next().unwrap();
+            if ids.insert(id.to_owned(), 0).is_some() {
+                eprintln!(
+                    "=> ID `{}` is defined more than once in the ID map in file `{}`",
+                    id, ID_MAP_PATH
+                );
+                *bad = true;
+            }
+        } else if line == "}" {
+            // We reached the end of the function.
+            break;
+        }
+    }
+    if ids.is_empty() {
+        eprintln!("=> No IDs were found in rustdoc in file `{}`...", ID_MAP_PATH);
+        *bad = true;
+    }
+    ids
+}
+
+fn check_id(
+    path: &Path,
+    id: &str,
+    ids: &mut HashMap<String, usize>,
+    line_nb: usize,
+    bad: &mut bool,
+) {
+    if id.contains('{') {
+        // This is a formatted ID, no need to check it!
+        return;
+    }
+    let id = id.to_owned();
+    match ids.get_mut(&id) {
+        Some(nb) => *nb += 1,
+        None => {
+            eprintln!(
+                "=> ID `{}` in file `{}` at line {} is missing from `init_id_map`",
+                id,
+                path.display(),
+                line_nb + 1,
+            );
+            *bad = true;
+        }
+    }
+}
+
+fn check_ids(
+    path: &Path,
+    f: &str,
+    ids: &mut HashMap<String, usize>,
+    regex: &Regex,
+    bad: &mut bool,
+) {
+    let mut is_checking_small_section_header = None;
+
+    for (line_nb, line) in f.lines().enumerate() {
+        let trimmed = line.trim_start();
+        // We're not interested in comments or doc comments.
+        if trimmed.starts_with("//") {
+            continue;
+        } else if let Some(start_line) = is_checking_small_section_header {
+            if line_nb == start_line + 2 {
+                check_id(path, trimmed.split('"').skip(1).next().unwrap(), ids, line_nb, bad);
+                is_checking_small_section_header = None;
+            }
+        } else if trimmed.starts_with("write_small_section_header(") {
```

Comment:
Similarly, it is currently too easy for this function to be renamed or argument order changed -- there's no comments that I can see indicating it's special. I think we should at least have comments to the definition. But still, parsing arbitrary Rust code seems worrisome (easy to get wrong).

Reply:
@GuillaumeGomez Could you add a doc-comment to `write_small_section_header()` saying that calls to it are parsed by tidy?

Also, this check seems somewhat fragile since it will only check calls that don't have any non-whitespace characters before them. Could you instead use something like this?

```suggestion
        } else if trimmed.contains("write_small_section_header(") && !trimmed.contains("fn write_small_section_header(") {
```

Reply:
Adding the doc comment and updating the check.